particle-iot/device-os
# Feature/hal/nfc

### Problem

Implement NFC Type 2 HAL driver

_NOTE: This PR changed the timer used by USART, because NFC uses timer 4. Now we use timer 1 in USART driver._

### Steps to Test

1. Flash example application to Xenon
2. Connect NFC antenna to Xenon
3. Use mobile phone NFC app to scan NFC antenna
4. When connected, D7 LED will turn on and you'll see the Particle website on your phone NFC app

### Example App

```c
#include "application.h"
#include "nfc_hal.h"
#include "nfc_ndef_msg.h"
#include "nfc_text_rec.h"
#include "nfc_uri_msg.h"
#include "nfc_t2t_lib.h"

SYSTEM_MODE(MANUAL);

static void nfc_event_callback(nfc_event_type_t type, nfc_event_t *event, void *ctx) {
    switch (type) {
        case NFC_EVENT_FIELD_ON: {
            digitalWrite(D7, 1);
            break;
        }
        case NFC_EVENT_FIELD_OFF: {
            digitalWrite(D7, 0);
            break;
        }
        case NFC_EVENT_READ: {
            break;
        }
        default:
            break;
    }
}

void setup(void) {
    static const uint8_t m_url[] = {'p', 'a', 'r', 't', 'i', 'c', 'l', 'e', '.', 'i', 'o'}; //URL "particle.com"
    static uint8_t m_ndef_msg_buf[256];

    // Just for NFC HAL API test, encode work will be done in wiring API
    uint32_t len = sizeof(m_ndef_msg_buf);
    uint32_t err_code = nfc_uri_msg_encode( NFC_URI_HTTP_WWW, m_url, sizeof(m_url), m_ndef_msg_buf, &len);
    APP_ERROR_CHECK(err_code);

    pinMode(D7, OUTPUT);
    digitalWrite(D7, 0);
    hal_nfc_type2_init();
    hal_nfc_type2_set_callback(nfc_event_callback);
    hal_nfc_type2_set_payload(m_ndef_msg_buf, len);
    hal_nfc_type2_start_emulation();
}

void loop() {

}
```

## Update

### NFC Wiring API Test
1. Flash  Example App
2. Use **NFC Reader** application on iOS to scan NFC Tag, you'll see three record
3. Click mode button, NFC records are updated

### Example App
```cpp
void nfc_event_handler(nfc_event_type_t type, nfc_event_t *event, void* ctx) {
    switch (type) {
        case NFC_EVENT_FIELD_ON:  digitalWrite(D7, 1); break;
        case NFC_EVENT_FIELD_OFF: digitalWrite(D7, 0); break;
        case NFC_EVENT_READ:      digitalWrite(D7, 0); break;
        default: break;
    }
}

void setup() {
    pinMode(D7, OUTPUT);
    digitalWrite(D7, 0);

    System.on(button_click, [](system_event_t ev, int clicks)->void {
        NFC.getNdefMessage()[0] = std::make_shared<TextRecord>("Update record - 1!", "en");
        NFC.getNdefMessage()[1] = std::make_shared<TextRecord>("Update record - 2!", "en");
        NFC.getNdefMessage()[2] = std::make_shared<TextRecord>("Update record - 3!", "en");
        NFC.getNdefMessage().addTextRecord("New record!", "en");

        NFC.stop();
        NFC.start(nfc_event_handler);
    });

    NFC.getNdefMessage().addTextRecord("Hello Particle!", "en");
    NFC.getNdefMessage().addUriRecord("particle.io", UriRecord::NFC_URI_HTTPS_WWW);
    NFC.getNdefMessage().addLauchAppRecord("io.particle.android.app");

    NFC.start(nfc_event_handler);
}

void loop() {

}
```
### Note
1. UID setting is not supported by Nordic Library, will be supported in the future
2. Lauch app only supports Android platform
3. TODO: Only support one message on NFC tag for now


---

### References

- [CH23817]
- https://github.com/particle-iot/nrf5_sdk/pull/4

---

### Completeness


---

- [gen 3] NFC support [#1606](https://github.com/particle-iot/device-os/pull/1606)
